### DISCUSSION TITLE
How to minimize the space inside rows?

### DISCUSSION THREAD

**Discussion by Yoyoda75** [Textualize/rich#3251] [General] [answered]
Hi everyone,

Here is a brief summary of my issue. I am trying to create Conway's game of life and my goal is to display the board in the terminal using Rich (maybe not the most efficient way, but it teaches me how to use Rich). The cells are filled with "■" when they are alive, and "□" when they are dead, so they are all the same width and the board should be a square. As you can see in the following screenshot, I managed to minimize the space between each *column*, but there is still space between each *row*:
 
![image](https://github.com/Textualize/rich/assets/60847050/0bdb2f62-c2d8-4c8a-8d79-717d8b3f26fb)

Here is the code that does it : 
```python
def update_table(board: Board) -> Table:
    table = Table(title="Game of Life", show_header=False, show_footer=False, show_edge=False, box=None)
    table.padding = (0, 0)
    table.pad_edge = False

    columns = board.shape[1]
    for column in range(columns):
        table.add_column(no_wrap=True)

    for row in board:
        formatted_row = ["■" if element else "□" for element in row]
        table.add_row(*formatted_row)

    return table
```

I have tried __*many*__ different combinations of options for Table(), but none let's me do what I want. So does anyone know how I could remove the space between each row entirely ? 

Here is the repo if you want to check out the full code (it's only two files): [Yoyoda75/game_of_life](https://github.com/Yoyoda75/game_of_life)

Thanks in advance !

**Comment by TomJGooding** [comment, marked as answer]
Unicode has a full block character (u2588) which hopefully will help explain the gaps between your rows. Try running the code below:

```python3
BLACK_SQUARE = "■"  # u25a0
FULL_BLOCK = "█"  # u2588

for _ in range(3):
    print(BLACK_SQUARE, FULL_BLOCK)
```

**Reply by Yoyoda75** [reply]
Thanks for your answer, but I don't think this is the result I am aiming for. 
![image](https://github.com/Textualize/rich/assets/60847050/ac24d492-63e8-406e-a6e6-9885393b8d3c)

I also tried changing the dead cells blocks to another full height block but the result is still not great:
![image](https://github.com/Textualize/rich/assets/60847050/ac79f034-8bdd-4eb7-91b1-d03abbcebb4b)

Since I was using square characters and the size of the board is x by x (say 50 by 50), I was planning on having a square board. However, the full height blocks are rectangular and make the board look distorted.

**Reply by Yoyoda75** [reply]
Never mind, I put two side by side and it seems to do the job.😄 It simply makes the cells quit big. Thank !

**Reply by TomJGooding** [reply]
I was trying to demonstrate that these gaps are actually in your Unicode square characters, so there isn't anything Rich can do about that.

Character cells in the terminal are roughly twice as high as they are wide, so one option would be to print two of the full block characters for a square:

```python
for _ in range(4):
    print("  ██  ██  ██  ██")
    print("██  ██  ██  ██  ")
```

Another option which would be more complicated but take up less space would be to use other [Unicode blocks](https://en.wikipedia.org/wiki/Block_Elements) - here's an example using upper/lower half blocks:

```python
for _ in range(4):
    print("▄▀▄▀▄▀▄▀")
```

[EDIT: Sorry I posted my message before I saw your reply that you'd figured the first part out!]
